### ISSUE TITLE
pmap error: "No canonicalize handler registered for type: <class 'tuple'>"

### ISSUE THREAD

**Issue by shoyer** [google/jax#1067]
@mattjj I can work up full code to reproduce, but just in case the traceback alone if enough for you...
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-d10a0485a179> in <module>()
----> 1 get_ipython().run_cell_magic('time', '', 'u_final, _ = multi_step_pmap((u, v), count=2**4, c=c, dt=0.5)\nu_final.block_until_ready()')

18 frames
google3/third_party/py/IPython/core/interactiveshell.py in run_cell_magic(self, magic_name, line, cell)
   2179             magic_arg_s = self.var_expand(line, stack_depth)
   2180             with self.builtin_trap:
-> 2181                 result = fn(magic_arg_s, cell)
   2182             return result
   2183 

google3/research/colab/lib/_debugger.py in time(self, line, cell, local_ns)
     92       _add_to_linecache('<timed exec>', code)
     93       _add_to_linecache('<timed eval>', code)
---> 94     return super(ColabExecutionMagics, self).time(line, cell, local_ns)
     95 
     96   def _debug_post_mortem(self):

<decorator-gen-60> in time(self, line, cell, local_ns)

google3/third_party/py/IPython/core/magic.py in <lambda>(f, *a, **k)
    191     # but it's overkill for just that one bit of state.
    192     def magic_deco(arg):
--> 193         call = lambda f, *a, **k: f(*a, **k)
    194 
    195         if callable(arg):

google3/third_party/py/IPython/core/magics/execution.py in time(self, line, cell, local_ns)
   1134         else:
   1135             st = clock2()
-> 1136             exec(code, glob, local_ns)
   1137             end = clock2()
   1138             out = None

<timed exec> in <module>()
----> 1 u_final, _ = multi_step_pmap((u, v), count=2**4, c=c, dt=0.5)
      2 u_final.block_until_ready()

<ipython-input-14-88c248ed1829> in multi_step_pmap(state, count, dt, c, exchange_interval)
    129     return lax.fori_loop(0, stop, lambda i, s: exchange_and_multi_step(s), state)
    130 
--> 131   return time_evolution(state)
    132 
    133   # evolved = time_evolution(sharded_state)

<ipython-input-14-88c248ed1829> in wrapper(state)
     99         lambda x: x.reshape((device_count, x.shape[0]//device_count) + x.shape[1:]),
    100         state)
--> 101     sharded_result = func(sharded_state)
    102     result = tree_util.tree_map(
    103         lambda x: x.reshape((-1,) + x.shape[2:]), sharded_result)

<ipython-input-14-88c248ed1829> in time_evolution(state)
    127   def time_evolution(state):
    128     stop = count // exchange_interval
--> 129     return lax.fori_loop(0, stop, lambda i, s: exchange_and_multi_step(s), state)
    130 
    131   return time_evolution(state)

google3/third_party/py/jax/lax/lax_control_flow.py in fori_loop(lower, upper, body_fun, init_val)
     84     return lax.add(i, lax._const(i, 1)), body_fun(i, x)
     85 
---> 86   _, result = while_loop(while_cond_fun, while_body_fun, (lower, init_val))
     87   return result
     88 

google3/third_party/py/jax/lax/lax_control_flow.py in while_loop(cond_fun, body_fun, init_val)
    153   out_flat = while_p.bind(
    154       init_val_flat, core.pack(cond_consts), core.pack(body_consts),
--> 155       aval_out=carry_aval_out, cond_jaxpr=cond_jaxpr, body_jaxpr=body_jaxpr)
    156   return build_tree(out_tree(), out_flat)
    157 

google3/third_party/py/jax/core.py in bind(self, *args, **kwargs)
    142     top_trace = find_top_trace(args)
    143     if top_trace is None:
--> 144       return self.impl(*args, **kwargs)
    145 
    146     tracers = map(top_trace.full_raise, args)

google3/third_party/py/jax/lax/lax_control_flow.py in _while_loop_impl(init_val, cond_consts, body_consts, aval_out, cond_jaxpr, body_jaxpr)
    164   val = init_val
    165   while cond_fun(val):
--> 166     val = body_fun(val)
    167   return val
    168 

google3/third_party/py/jax/core.py in eval_jaxpr(jaxpr, consts, freevar_vals, *args)
    198                in eqn.bound_subjaxprs]
    199     subfuns = map(lu.wrap_init, subfuns)
--> 200     ans = eqn.primitive.bind(*(subfuns + in_vals), **eqn.params)
    201     outvals = list(ans) if eqn.destructure else [ans]
    202     map(write, eqn.outvars, outvals)

google3/third_party/py/jax/core.py in call_bind(primitive, f, *args, **params)
    661   if top_trace is None:
    662     with new_sublevel():
--> 663       ans = primitive.impl(f, *args, **params)
    664   else:
    665     tracers = map(top_trace.full_raise, args)

google3/third_party/py/jax/interpreters/pxla.py in xla_pmap_impl(fun, *args, **params)
    509   abstract_args = map(partial(abstractify, axis_size), args)
    510   compiled_fun = parallel_callable(fun, axis_name, axis_size, *abstract_args)
--> 511   return compiled_fun(*args)
    512 
    513 def abstractify(axis_size, x):

google3/third_party/py/jax/interpreters/pxla.py in execute_replicated(compiled, pval, nrep, handle_in, handle_replica_result, handle_full_result, *args)
    572            "XLA devices are available")
    573     raise ValueError(msg.format(nrep, xb.device_count()))
--> 574   input_bufs = zip(*map(handle_in, args)) if args else [[]] * nrep
    575   out_bufs = compiled.ExecutePerReplica(list(input_bufs))
    576   results = [merge_pvals(handle_replica_result(buf), pval) for buf in out_bufs]

google3/third_party/py/jax/interpreters/pxla.py in shard_arg(device_ordinals, axis_size, arg)
     69     return [buf if buf.device() == device_ordinals[r]
     70             else xla.device_put(get_shard(i), device_ordinals[r])
---> 71             for r, (i, buf) in enumerate(zip(assignments, arg.device_buffers))]
     72   else:
     73     shards = [(_slice(arg, assignments[i]), device_ordinals[i])

google3/third_party/py/jax/interpreters/pxla.py in <listcomp>(.0)
     69     return [buf if buf.device() == device_ordinals[r]
     70             else xla.device_put(get_shard(i), device_ordinals[r])
---> 71             for r, (i, buf) in enumerate(zip(assignments, arg.device_buffers))]
     72   else:
     73     shards = [(_slice(arg, assignments[i]), device_ordinals[i])

google3/third_party/py/jax/interpreters/xla.py in device_put(x, device_num)
    140     A buffer representing the input `x` placed on the appropriate device.
    141   """
--> 142   x = _canonicalize_pyval_dtype(x)
    143   t = type(x)
    144   if t is DeviceArray or t is DeviceTuple:

google3/third_party/py/jax/interpreters/xla.py in _canonicalize_pyval_dtype(x)
    418   except KeyError:
    419     msg = "No canonicalize handler registered for type: {}"
--> 420     raise TypeError(msg.format(type(x)))
    421 
    422 canonicalize_dtype_handlers = {}

TypeError: No canonicalize handler registered for type: <class 'tuple'>
```

**Comment by shoyer** [comment]
To reproduce:
```python
import jax.numpy as np
import jax
from jax import lax
from functools import partial

from jax.lib import xla_bridge
from jax import tree_util
device_count = xla_bridge.device_count()


# patch https://github.com/google/jax/pull/1063
from jax.interpreters.pxla import ShardedDeviceTuple
xla_bridge.register_constant_handler(ShardedDeviceTuple, jax.xla._device_tuple_constant_handler)


@jax.jit
def multi_step(state, count):
  return lax.fori_loop(0, count, lambda i, s: s, state)


def multi_step_pmap(state, count, exchange_interval=1):

  @partial(jax.pmap, axis_name='x')
  @jax.jit
  def exchange_and_multi_step(state):
    return multi_step(state, exchange_interval)

  def time_evolution(state):
    stop = count // exchange_interval
    return lax.fori_loop(0, stop, lambda i, s: exchange_and_multi_step(s), state)
  
  return time_evolution(state)

u = np.zeros((device_count, 128))
u_final = multi_step_pmap(u, count=2**3)
```

It seems to be another nested jit issue not unlike https://github.com/google/jax/issues/1065 -- removing the inner jit removes the error.

**Comment by mattjj** [comment]
We're currently attempting to remove all tuples from the system (an internals-only change), which would make this error moot. Since Stephan said this isn't blocking his work, I'm going to put off thinking about this until we see if we can remove tuples entirely!

**Comment by mattjj** [comment]
Seems fixed by #1224 ! Thanks, @joaogui1 .
